base/generator.jl

```diff
@@ -126,3 +132,9 @@ IteratorEltype(::Type) = HasEltype()  # HasEltype is the default
 IteratorEltype(::Type{Generator{I,T}}) where {I,T} = EltypeUnknown()
 
 IteratorEltype(::Type{Any}) = EltypeUnknown()
+
+# See comment for iteratorsize re invalidation
+function iteratoreltype(::Type{T}) where T
+    T<:(Tuple{Vararg{X}} where X) ? HasEltype() :
+        T<:Tuple ? EltypeUnknown() : IteratorEltype(T)
```

Comment:
This definition makes sense, but currently we always give `HasEltype` for tuples (for better or worse).